core/src/test/java/kafka/log/remote/RemoteLogManagerInteractionTest.java

```diff
@@ -0,0 +1,323 @@
+        // uploaded multiple times before succeeding again.
+        // At the end we should see no call to delete as retention.bytes was never breached
+        RemoteLogManager.RLMTask copyTask = remoteLogManager.new RLMCopyTask(topicIdPartition, 128);
+        RemoteLogManager.RLMTask cleanupTask = remoteLogManager.new RLMExpirationTask(topicIdPartition);
+
+        buildSegment(retentionSize /  3);
+        log.maybeUpdateHighWatermark(log.logEndOffset());
+        when(remoteStorageManager.copyLogSegmentData(any(), any())).thenReturn(Optional.empty());
+        copyTask.run();
+
+        buildSegment(retentionSize /  3);
+        log.maybeUpdateHighWatermark(log.logEndOffset());
+        when(remoteStorageManager.copyLogSegmentData(any(), any())).thenThrow(new RemoteStorageException(""));
+        for (int i = 0; i < 4; i++) {
+            copyTask.run();
+        }
+        doReturn(Optional.empty()).when(remoteStorageManager).copyLogSegmentData(any(), any());
+        copyTask.run();
+
+        buildSegment(retentionSize /  3);
+        log.maybeUpdateHighWatermark(log.logEndOffset());
+        copyTask.run();
+        cleanupTask.run();
+
+        // total is 1 (success) + 4 (failures) + 2x1 (success)
+        verify(remoteStorageManager, times(7)).copyLogSegmentData(any(), any());
+        verify(remoteStorageManager, never()).deleteLogSegmentData(
+            argThat(segment -> segment.state().equals(RemoteLogSegmentState.COPY_SEGMENT_FINISHED))
+        );
+        verify(remoteStorageManager, times(4)).deleteLogSegmentData(
+            argThat(segment -> segment.state().equals(RemoteLogSegmentState.COPY_SEGMENT_STARTED))
+        );
+    }
+
+    @Test
+    void failedSegmentsAreNotAccountedInSizeDuringDeletion() throws RemoteStorageException {
+        RemoteLogManager.RLMTask copyTask = remoteLogManager.new RLMCopyTask(topicIdPartition, 128);
+        RemoteLogManager.RLMTask cleanupTask = remoteLogManager.new RLMExpirationTask(topicIdPartition);
+
+        buildSegment(retentionSize - 100);
+        log.maybeUpdateHighWatermark(log.logEndOffset());
+        when(remoteStorageManager.copyLogSegmentData(any(), any())).thenThrow(new RemoteStorageException(""));
+        for (int i = 0; i < 4; i++) {
+            copyTask.run();
+        }
+        doReturn(Optional.empty()).when(remoteStorageManager).copyLogSegmentData(any(), any());
+        copyTask.run();
+
+        buildSegment(retentionSize - 100);
+        log.maybeUpdateHighWatermark(log.logEndOffset());
+        copyTask.run();
+        cleanupTask.run();
+
+        // total is 1 (success) + 4 (failures) + 1 (success)
+        verify(remoteStorageManager, times(6)).copyLogSegmentData(any(), any());
+
+        // The first segment should now be deleted
+        verify(remoteStorageManager, times(5)).deleteLogSegmentData(
+            argThat(segment -> segment.startOffset() == 0)
+        );
```

Comment:
Could you help me understand why we call 5 times of deleteLogSegmentData here? Maybe add the explanation as a comment above the line. Thanks.

Reply:
```
        when(remoteStorageManager.copyLogSegmentData(any(), any())).thenThrow(new RemoteStorageException(""));
         for (int i = 0; i < 4; i++) {
             copyTask.run();
         }
```
This portion will have created 4 failed `COPY_SEGMENT_STARTED` segment and we have a successful run after. So in total we have 5 segments metadata referring to the first segment.
All those segments are meant for deletion due to `retention.bytes` policy as we pushed a second segment after it, which is what is causing the 4 failures + 1 success to be deleted.